mount.c

```diff
@@ -136,8 +150,23 @@ static void do_mount(const char *source, const char *target, const char *type, u
 
 	int ro_bind = (flags & (MS_BIND | MS_RDONLY)) == (MS_BIND | MS_RDONLY)
 			&& !(flags & MS_REMOUNT);
-	if (ro_bind && mount("none", target, NULL, flags | MS_REMOUNT, NULL) == -1) {
-		err(1, "mount_entries: read-only remount of %s", target);
+	if (ro_bind) {
+		int rc;
+#ifdef HAVE_SYS_mount_setattr
+		unsigned int aflags = 0;
+		if (flags & MS_REC) {
+			aflags |= AT_RECURSIVE;
+		}
+		struct mount_attr attr = {
+			.attr_set = MOUNT_ATTR_RDONLY,
+		};
+		rc = mount_setattr(AT_FDCWD, target, aflags, &attr, sizeof (attr));
+#else
+		rc = mount("none", target, NULL, flags | MS_REMOUNT, NULL);
```

Comment:
Instead of #else, how about falling back to mount(2) if mount_setattr(2) is unavailable at runtime?